tests/timeseries/test_quantities.py

```diff
@@ -531,3 +537,55 @@ def test_invalid_multiplications() -> None:
             _ = quantity * 200.0  # type: ignore
         with pytest.raises(TypeError):
             quantity *= 200.0  # type: ignore
+
+
+def test_to_and_from_string2() -> None:
+    """Test string parsing and formatting."""
+    random.seed(0)
+    exponents_to_test = [0, 3, 6, 9]
+
+    for exponent in exponents_to_test:
+        for _ in range(10):
+            value = random.random() * 10**exponent
+
+            power_str = f"{Power.from_watts(value):.{exponent}}"
+            assert f"{Power.from_string(power_str):.{exponent}}" == power_str
+            print(power_str)
+
+            voltage_str = f"{Voltage.from_volts(value):.{exponent}}"
+            assert f"{Voltage.from_string(voltage_str):.{exponent}}" == voltage_str
+            print(voltage_str)
+
+            current_str = f"{Current.from_amperes(value):.{exponent}}"
+            assert f"{Current.from_string(current_str):.{exponent}}" == current_str
+
+            energy_str = f"{Energy.from_watt_hours(value):.{exponent}}"
+            assert f"{Energy.from_string(energy_str):.{exponent}}" == energy_str
+
+            freq_str = f"{Frequency.from_hertz(value):.{exponent}}"
+            assert f"{Frequency.from_string(freq_str):.{exponent}}" == freq_str
+
+            pct_str = f"{Percentage.from_fraction(value):.{exponent}}"
+            assert f"{Percentage.from_string(pct_str):.{exponent}}" == pct_str
+
+
+@pytest.mark.parametrize("quantity_type", [Power, Voltage, Current, Energy, Frequency])
+@pytest.mark.parametrize("exponent", [0, 3, 6, 9])
+@hypothesis.settings(max_examples=1000)
+@hypothesis.seed(42)  # Good seed that triggers a lot of problematic edge cases
+@hypothesis.given(value=st.floats(min_value=-1.0, max_value=1.0))
+def test_to_and_from_string(
+    quantity_type: type[Quantity], exponent: int, value: float
+) -> None:
+    """Test string parsing and formatting."""
+    quantity = quantity_type.__new__(quantity_type)
+    quantity._base_value = value * 10**exponent
+    quantity_str = f"{quantity:.{exponent}}"
+    from_string = quantity_type.from_string(quantity_str)
+    try:
+        assert f"{from_string:.{exponent}}" == quantity_str
+    except AssertionError as error:
+        pytest.fail(
+            f"Failed for {quantity.base_value} != from_string({from_string.base_value}) "
+            + f"with exponent {exponent} and source value '{value}': {error}"
+        )
```

Comment:
Is this just to make debugging easier? Maybe an alternative would be to keep the normal assert and before asserting (always) printing the extra context info, like `print(f"{quantity.base_value=}, {from_string.base_value=} ...", )`. This way when any test fails you'll have all the info you need in the test failure output.

But it is just a suggestion, I'm OK with leaving it as it is now.

Reply:
Yeah, it was mainly for debugging